---
title: 'Microtus oregoni Autosomal vs Sex chromosome genetic structure '
author: "Bonginkosi Charles Gumbi"
date: '`r Sys.Date()`'
output:
  pdf_document: default
  html_document: default
email: bonginkg@ucr.edu
---

installing packages if running the script for the first time on your computer

```{r install packages}
# Installing from GitHub requires installing devtools
#packages("devtools")
#devtools::install_github("SexGenomicsToolkit/sgtr")
#devtools::install_github("TheWangLab/algatr")
#install.packages("parallelly", dependencies = T)

# Install the r2vcftools library from GitHub
#devtools::install_github("Bioconductor-mirror/LEA", force=T)
#devtools::install_github("nspope/r2vcftools", force=T)
#remotes::install_github("nspope/r2vcftools")
#if (!requireNamespace("BiocManager", quietly=TRUE))
#    install.packages("BiocManager")
#BiocManager::install("gdsfmt")
#BiocManager::install("SNPRelate")
```

loading packages

```{r loading libraries}
library(sgtr)
library(stringr)
library(adegenet)
library(data.table)
library(vcfR)
library(RColorBrewer)
library(reshape)
library(hierfstat)
library(tidyverse)
library(pcadapt)
library(tibble)
library(here)
library(pinfsc50)
library(utils)
library(fields)
library(admisc)
library(vegan)
library(lfmm)
library(spdep)
library(LEA)
library(wordcloud)
library(ggmap)
library(maps)
library(mapdata)
library(stringr)
library(sp)
library(sf)
library(algatr)
library(here)
library(viridis)
library(tidyr)
library(gridExtra)
library(tibble)
library(SNPRelate)
library(SeqArray)
library(wingen)
library(purrr)
library(cowplot)
library(vcfR)
library(dplyr)
library(gdsfmt)
library(SNPRelate)
library(tidyverse)
library(tidyverse)
library(readr)
library(ggplot2)
library(forcats)
library(ggthemes)
library(patchwork)
```

removing objects from the environment

```{r clear the enviroment}
rm(list=ls())
```

loading genetic data and filtering

```{r datasets}
#SNPs dataset
nohapvcf <- read.vcfR("~/MIOR/moregoni/data/nohap.populations.snps.vcf", convertNA = TRUE, skip = 0, checkFile = T, check_keys = T)
#metadata dataset
metadata <- read.csv ("~/MIOR/moregoni/data/mior_metadata_20250420.csv", header = T)
```
converting vcf file to genind object where population is define by location

```{r  location vcf to genind, echo=FALSE}
set.seed(101)
nohapgenind <- vcfR2genind(nohapvcf)
nohapgenind_df <- adegenet::genind2df(nohapgenind, usepop = F)
nohapgenind_df <- tibble::rownames_to_column(nohapgenind_df, "sample.id")
nohapgenind_df <- merge(nohapgenind_df, metadata, by = "sample.id")

# Adding metadata into genind
other_data <- subset(nohapgenind_df[ , -(2:225147 )])

# Adding gps points into the genind 
x <- c(other_data$lat)
y <-  c(other_data$long)
coord <- data.frame(x, y)

nohapgenind <- adegenet::df2genind(X = nohapgenind_df[,c(2:225147)], sep="", ncode = 1, 
                             ind.names = nohapgenind_df$sample.id, loc.names = NULL, 
                             pop = nohapgenind_df$location, NA.char = "NA", ploidy = 2, 
                             type="codom", strata = other_data, hierarchy=NULL)
pop(nohapgenind)
```

sample size summary population

```{r sample size }
 nohapgenind_df %>%
  group_by(location)  %>%
  count()
```


population structure analysis using ADMIXTURE output

```{r}
samplelist <- read_tsv("~/MIOR/moregoni/pop_structure/file.list", 
                       col_names = "sample.id")

all_data <- tibble(sample.id=character(),
                   k=numeric(),
                   Q=character(),
                   value=numeric())

for (k in 1:10){
  data <- read_delim(paste0("~/MIOR/moregoni/pop_structure/file.",k,".Q"),
                     col_names = paste0("Q",seq(1:k)),
                     delim=" ")
  data$sample.id <- samplelist$sample.id
  data$k <- k
  
  #This step converts from wide to long.
  data %>% gather(Q, value, -sample.id,-k) -> data
  all_data <- rbind(all_data,data)
}

all_data <- merge(all_data, metadata, by = "sample.id")

k2plot <- all_data %>%
   filter(k == 2) %>%
  ggplot(aes(factor(sample.id), value, fill = factor(Q))) +
  geom_col(color = "gray", size = 0.1) +
  facet_grid(~fct_inorder(location), switch = "x", scales = "free", space = "free") +
  theme_minimal() + labs(x = "Individuals", title = "K=2", y = "Ancestry") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_discrete(expand = expand_scale(add = 1)) +
  theme(panel.spacing.x = unit(0.1, "lines"),
    axis.text.x = element_blank(),
    panel.grid = element_blank()) +
  scale_fill_gdocs(guide = FALSE)

k3plot <- all_data %>%
   filter(k == 3) %>%
  ggplot(aes(factor(sample.id), value, fill = factor(Q))) +
  geom_col(color = "gray", size = 0.1) +
  facet_grid(~fct_inorder(location), switch = "x", scales = "free", space = "free") +
  theme_minimal() + labs(x = "Individuals", title = "K=3", y = "Ancestry") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_discrete(expand = expand_scale(add = 1)) +
  theme(panel.spacing.x = unit(0.1, "lines"),
    axis.text.x = element_blank(),
    panel.grid = element_blank()) +
  scale_fill_gdocs(guide = FALSE)

k4plot <- all_data %>%
   filter(k == 4) %>%
  ggplot(aes(factor(sample.id), value, fill = factor(Q))) +
  geom_col(color = "gray", size = 0.1) +
  facet_grid(~fct_inorder(location), switch = "x", scales = "free", space = "free") +
  theme_minimal() + labs(x = "Individuals", title = "K=4", y = "Ancestry") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_discrete(expand = expand_scale(add = 1)) +
  theme(panel.spacing.x = unit(0.1, "lines"),
    axis.text.x = element_blank(),
    panel.grid = element_blank()) +
  scale_fill_gdocs(guide = FALSE)

combined_k = k2plot + k3plot + k4plot + plot_layout(ncol = 1)

ggsave(plot = k2plot, filename = "~/MIOR/moregoni/out_plots/K2_plot.png", width =8, height =6, dpi = 300)
ggsave(plot = k3plot, filename = "~/MIOR/moregoni/out_plots/K3_plot.png", width =8, height =6, dpi = 300)
ggsave(plot = k4plot, filename = "~/MIOR/moregoni/out_plots/K4_plot.png", width =8, height =6, dpi = 300)
ggsave(plot = combined_k, filename = "~/MIOR/moregoni/out_plots/combined_k_plot.png", width =6, height =8, dpi = 300)
```


Principal Component Analysis with SNPRelate

```{r}
library(SNPRelate)
library(tidyverse)
snpgdsVCF2GDS("~/MIOR/moregoni/data/nohap.populations.snps.vcf",
              "~/MIOR/moregoni/data/nohap_population.filtered.gds", method="biallelic.only")
# lodianggds object 
genofile <- openfn.gds("~/MIOR/moregoni/data/nohap_population.filtered.gds", readonly = F)
```

adding data to gds object 

```{r}
add.gdsn(genofile, "pop.group", metadata$location)
add.gdsn(genofile, "sex", metadata$sex)
add.gdsn(genofile, "state", metadata$state)

```

pruning linked loci

```{r}
snpset_pruned <- snpgdsLDpruning(genofile, autosome.only = F)
names(snpset_pruned)

# Making a list of sites 
snpset.id <- unlist(snpset_pruned)
# getting  all selected snp id
snpset.id <- unlist(unname(snpset_pruned))
head(snpset.id)
```

running PCA

```{r}
pca <- snpgdsPCA(genofile, num.thread = 2, eigen.cnt = 16, snp.id = snpset.id, 
                 missing.rate = 0.10, maf = 0.05, autosome.only = F)

# principal Component Analysis (PCA) on genotypes:
pca

# calculating percent variance explained for each eigenvector
pc.percent <- pca$varprop*100
round(pc.percent, 2)


# making a data frame of your PCA results
tab <- data.frame(sample.id = pca$sample.id,
                  PC1 = pca$eigenvect[,1],    # the first eigenvector
                  PC2 = pca$eigenvect[,2],    # the second eigenvector
                  stringsAsFactors = FALSE)


tab <- merge(tab, metadata, by = "sample.id")

# ploting pallete
q_palette <- c("#440154", "#0072B2", "#D55E00")

pca_plot <- tab %>%
  ggplot(.,aes(x = PC1, y = PC2)) +
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = 0) + 
  geom_point(aes(colour = location, shape = sex), 
             size = 3, show.legend = T) + # centroids
  scale_colour_manual(values = q_palette) +
  xlim(-0.3, 0.3) +  # Set x-axis limits
  ylim(-0.3, 0.3) +  # Set y-axis limits
  labs(x = "Linear discriminant 1 (19.5%)", y = "Linear discriminant 2 (9.7%)", 
       shape = "Sex", colour = "Population") +
  theme_cowplot() +
  theme(axis.text.y = element_text(colour="black", size=12), # Custom theme for ggplot2
        axis.text.x = element_text(colour="black", size = 12),
        axis.title = element_text(colour="black", size = 12),
        panel.border = element_rect(colour="black", fill = NA, size=1),
        panel.background = element_blank(),
        plot.title = element_text(hjust=0.5, size=15),
        text = element_text(size = 12,  family = "sans"))

ggsave(plot = pca_plot, filename = "~/MIOR/moregoni/out_plots/pca_plot.png", width = 6, height = 8, dpi = 300)
```

